content/algorithms/expected_consensus/_index.md

```diff
@@ -28,115 +28,290 @@ The [Storage Power Consensus](storage_power_consensus) subsystem uses access to
 
-We show this below, removing division for ease of implementation:
+- The power table at epoch `X` shows that the attacker has sectors 1, 2, 3, 4, 5.
+- The miner decides to not store sector 5.
+- The miner wins the election at epoch `X`.
+  - Main fork: Miner is asked a WinningPoSt for sector 5 for which they won't be able to provide a proof.
+  - The miner creates a fork, terminates sector 5 in epochs before `X`.
+  - At `X`, the miner is now challenged a different sector (not 5).
 
-```text
-const maxChallengeTicketSize = 2^len(H)
+Note that there are variants of this attack in which the miner introduces a new sector to change which sector will be challenged.
+
+> What happens if a sector expired after `WinningPoStSectorSetLookback`?
+
+An expired sector will not be challenged during WindowPoSt (hence not penalized after its expiration). However, an edge case around the fact that `WinningPoStSectorSetLookback` is longer than finality is that due to the lookback, a miner can be challenged for an expired sector between `expirationEpoch` and `expirationEpoch + WinningPoStSectorSetLookback - 1`. Therefore, it is important that miners keep an expired sector for `WinningPoStSectorSetLookback` more epochs after expiration, or they will not be able to generate a WinningPoSt (and get the corresponding reward).
+
+Example:
+- At epoch `X`:
+  - Sector expires and miner deletes the sector.
+- At epoch `X+WinningPoStSectorSetLookback-1`:
+  - The expired sector gets selected for WinningPoSt
+  - The miner will not be able to generate the WinningPoSt and they will not win the sector.
+
+**Step 3:** Block creation and propagation
+
+If the above is successful, miners build a block and propagate it.
+
+### GenerateElectionProof
+
+`GenerateElectionProof` outputs the result of whether a miner won the block or not as well as the quality of the block mined.
+
+The "Block Quality" is an integer that is used for weight and block reward calculations. For example a block with quality "2" is equivalent as two blocks of quality "1".
+
+#### High level algorithm
+
+- Get the percentage of power at block `ElectionPowerTableLookback`
+  - Get the power of the miner at block `ElectionPowerTableLookback`
+  - Get the total network power at block `ElectionPowerTableLookback`
+- Get randomness for the current epoch using `GetRandomness`.
+- Generate a VRF and compute its hash
+  - The storage miner's `workerKey` is given as input in the VRF process
+- Compute block quality: The smaller the hash is, the higher the block quality will be
+  - Compute the probability distribution of winning `k` blocks (i.e. Poisson distribution see below for details)
+  - Let `h_n` be the normalised VRF, i.e. `h_n = vrfout/2^256`
+  - The probability of winning one block is `1-P[X=0]`, thus if `h_n` is less than `1-P[X=0]`, the miner wins at least one block.
+  - Similarly if `h_n` is less than `1-P[X=0]-P[X=1]` we have at least two blocks and so on.
```

Comment:
I might add: "While it is not permitted for a single miner to publish two distinct blocks, in this case, the miner produces a single block which earns two block rewards."